File: Bellman-Ford.cpp
#include <iostream>
#include <vector>
#include <limits>
#include <algorithm>

using namespace std;

struct Edge {
    int u, v, w;
};

class Graph {
public:
    Graph(int vertices);
    void addEdge(int u, int v, int w);
    bool bellmanFord(int src, int dest);

private:
    int V;
    vector<Edge> edges;
};

Graph::Graph(int vertices) {
    V = vertices;
}

void Graph::addEdge(int u, int v, int w) {
    edges.push_back({u, v, w});
}

bool Graph::bellmanFord(int src, int dest) {
    vector<int> dist(V, numeric_limits<int>::max());
    dist[src - 1] = 0;

    for (int i = 0; i < V - 1; i++) {
        for (const Edge &edge : edges) {
            int u = edge.u;
            int v = edge.v;
            int weight = edge.w;
            if (dist[u - 1] != numeric_limits<int>::max() && dist[u - 1] + weight < dist[v - 1]) {
                dist[v - 1] = dist[u - 1] + weight;
            }
        }
    }

    for (const Edge &edge : edges) {
        int u = edge.u;
        int v = edge.v;
        int weight = edge.w;
        if (dist[u - 1] != numeric_limits<int>::max() && dist[u - 1] + weight < dist[v - 1]) {
            return true; // Negative cycle detected
        }
    }

    if (dist[dest - 1] == numeric_limits<int>::max()) {
        cout << "No path from source to destination" << endl;
    } else {
        vector<int> path;
        int current = dest - 1;
        while (current != src - 1) {
            for (const Edge &edge : edges) {
                int u = edge.u;
                int v = edge.v;
                int weight = edge.w;
                if (dist[u - 1] + weight == dist[current]) {
                    path.push_back(current + 1);
                    current = u - 1;
                    break;
                }
            }
        }
        path.push_back(src);
        reverse(path.begin(), path.end());

        for (int node : path) {
            cout << node << " ";
        }
        cout << endl << dist[dest - 1] << endl;
    }

    return false; // No negative cycle detected
}

int main() {
    int n, e;
    cin >> n >> e;
    Graph g(n);

    for (int i = 0; i < e; i++) {
        int u, v, w;
        cin >> u >> v >> w;
        g.addEdge(u, v, w);
    }

    int src, dest;
    cin >> src >> dest;

    if (g.bellmanFord(src, dest)) {
        cout << "Negative Cycle Detected" << endl;
    }

    return 0;
}

File: Bellman-Ford.py
class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.graph = []

    def add_edge(self, u, v, w):
        self.graph.append([u, v, w])

    def bellman_ford(self, src, dest):
        dist = [float("inf")] * self.V
        dist[src - 1] = 0

        for _ in range(self.V - 1):
            for u, v, w in self.graph:
                if dist[u - 1] != float("inf") and dist[u - 1] + w < dist[v - 1]:
                    dist[v - 1] = dist[u - 1] + w

        # Check for negative cycles
        for u, v, w in self.graph:
            if dist[u - 1] != float("inf") and dist[u - 1] + w < dist[v - 1]:
                print("Negative Cycle Detected")
                return

        # Reconstruct the path if it exists
        path = [dest]
        current = dest - 1
        while current != src - 1:
            for u, v, w in self.graph:
                if dist[u - 1] + w == dist[current]:
                    path.append(u)
                    current = u - 1
                    break

        path.reverse()

        if dist[dest - 1] == float("inf"):
            print("No path from source to destination")
        else:
            print(" ".join(map(str, path)))
            print(dist[dest - 1])

if __name__ == "__main__":
    n = int(input())
    e = int(input())
    g = Graph(n)

    for _ in range(e):
        u, v, w = map(int, input().split())
        g.add_edge(u, v, w)

    src = int(input())
    dest = int(input())

    g.bellman_ford(src, dest)
